src/OrchardCore/OrchardCore.Media.Core/DefaultMediaFileStore.cs

```diff
@@ -169,24 +169,34 @@ public virtual async Task<string> CreateFileFromStreamAsync(string path, Stream
 
                 foreach (var mediaCreatingEventHandler in _mediaCreatingEventHandlers)
                 {
-                    // Creating stream disposed by using.
-                    using var creatingStream = outputStream;
-
-                    // Stop disposal of inputStream, as creating stream is the object to dispose.
-                    inputStream = null;
-
-                    // Outputstream must be created by event handler.
+                    var creatingStream = outputStream;                    
                     outputStream = null;
 
-                    outputStream = await mediaCreatingEventHandler.MediaCreatingAsync(context, creatingStream);
+                    try
+                    {
+                        outputStream = await mediaCreatingEventHandler.MediaCreatingAsync(context, creatingStream);
+                    }
+                    catch
+                    {
+                        outputStream = creatingStream;
+                    }
+                    finally
+                    {
+                        if (creatingStream != outputStream)
+                        {
+                            creatingStream.Dispose();
+                        }
+                    }
                 }
 
                 return await _fileStore.CreateFileFromStreamAsync(context.Path, outputStream, overwrite);
             }
             finally
             {
-                // This disposes the last outputStream.
-                outputStream?.Dispose();
+                if (inputStream != outputStream)
+                {
+                    inputStream = outputStream;                    
```

Comment:
You can't 'return' the stream this way. If you want to update the reference for the caller as well, you must pass the argument by reference, e.g., `CreateFileFromStreamAsync(string path, ref Stream inputStream, bool overwrite = false)`. However, this requires changes to the interface contract.